finality-aleph/src/network.rs

```diff
@@ -196,22 +209,29 @@ impl Peers {
                 self.to_peer.entry(*session_id).and_modify(|hm| {
                     hm.remove(node_id);
                 });
+                self.session_peers
+                    .entry(*session_id)
+                    .and_modify(|v| v.retain(|p| *p != *peer));
             }
         }
-        self.to_peer.retain(|_, hm| !hm.is_empty());
+        self.session_peers.retain(|_, v| !v.is_empty());
     }
 
     fn peers_authenticated_for(&self, session_id: SessionId) -> impl Iterator<Item = &PeerId> {
-        self.to_peer
-            .get(&session_id)
-            .into_iter()
-            .map(|hm| hm.values())
-            .flatten()
+        self.session_peers.get(&session_id).into_iter().flatten()
     }
 
     fn get(&self, session_id: SessionId, node_id: NodeIndex) -> Option<&PeerId> {
         self.to_peer.get(&session_id)?.get(&node_id)
     }
+
+    fn get_rand(&self, session_id: SessionId, n_peers: usize) -> impl Iterator<Item = &PeerId> {
+        self.session_peers
+            .get(&session_id)
+            .expect("the session already started")
+            .as_slice()
+            .choose_multiple(&mut rand::thread_rng(), n_peers)
```

Comment:
I'm a bit concerned about the `thread_rng` here, this is usually quite a slow operation and sending messages is a very common occurrence. Perhaps instead of making this random do a simple round-robin on the set of peers to send to? Or at least keep the thread RNG somewhere so that further calls are only pseudorandom?

Reply:
Added a pseudorandom method but I'm not entirely satisfied with it. I'll welcome any improvements.